### PR TITLE
[Auto Downloading] - Disable all podcasts auto download status when disabling the Global Auto Download

### PR DESCRIPTION
**Pull request by mebarbosa**
## Description
- This PR adds the ability back to disable all auto download podcast status after disabling the global Auto download flag
- Put back the code that was removed here: https://github.com/Automattic/pocket-casts-android/pull/3020/files
- To handle cases where users have already configured global auto-download, I refactored `autoDownloadNewEpisodes` to be a `UserSetting<Int>` instead of a `UserSetting<Boolean>`. This change allows me to determine whether the user manually disabled the `New Episodes` toggle or if it was simply unset due to using an app version prior to `7.76`, which does not yet include the autoDownloadNewEpisodes toggle
- This issue was reported here: pdeCcb-7pt-p2#comment-5894 in our Beta Testing

### Disable all podcasts auto download
1. Fresh install
2. Subscribe to a podcast
3. ✅ Verify the episodes are auto downloading
4. ✅ Verify the podcast has auto download enabled
5. Go to the Profile tab -> Settings cog -> Auto Download
6. Toggle off “New Episodes”
7. Go to a podcast settings page
8. ✅ Verify the auto download has been turned off for the podcast too

### App Update
1. Install the app from 946dda11c757ac87c7b4ca29dd3d8725482933b5
2. Subscribe to some podcasts and enable their auto download toggle
3. Open Profile -> Settings -> Auto Download
4. ✅ Ensure you have New Episodes toggle ON
5. Install the app from this branch
6. Open Profile -> Settings -> Auto Download
7. ✅ Ensure you have New Episodes toggle ON

## Screenshots or Screencast 
[Screen_recording_20241104_103007.webm](https://github.com/user-attachments/assets/a2e3ffa6-ac5f-45b5-975f-8c6c57128182)

## Checklist
- [ ] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [x] I have considered whether it makes sense to add tests for my changes
- [x] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [x] Any jetpack compose components I added or changed are covered by compose previews
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.
 
#### I have tested any UI changes...

- [x] with different themes
- [x] with a landscape orientation
- [x] with the device set to have a large display and font size
- [x] for accessibility with TalkBack

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/Automattic/pocket-casts-android/milestone/67">7.76 ❄️</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: modules/features/settings/src/main/java/au/com/shiftyjelly/pocketcasts/settings/AutoDownloadSettingsFragment.kt [hunk 1]
```diff
@@ -238,6 +238,9 @@ class AutoDownloadSettingsFragment :
 
     private fun handleNewEpisodesToggle(isEnabled: Boolean) {
         lifecycleScope.launch {
+            if (!isEnabled) {
+                async(Dispatchers.IO) { podcastManager.updateAllAutoDownloadStatus(Podcast.AUTO_DOWNLOAD_OFF) }.await()
```
Comment (geekygecko) [inline modules/features/settings/src/main/java/au/com/shiftyjelly/pocketcasts/settings/AutoDownloadSettingsFragment.kt:242]: This is the only place that calls `updateAllAutoDownloadStatus` so we should convert the DAO method to a suspend function and let Room handle the context.

**Comment by CookieyedCodes** [comment]
So when my app updated the new toggle had switched it off, luckily I caught it and turned my new episodes back on that saved my 400+ new episodes selected, so while yes I would like it to wipe, a backup or delay functionality would be appreciated just in case 😅

File: modules/features/settings/src/main/java/au/com/shiftyjelly/pocketcasts/settings/AutoDownloadSettingsFragment.kt (hunk 1 above)
Comment (mebarbosa) [reply modules/features/settings/src/main/java/au/com/shiftyjelly/pocketcasts/settings/AutoDownloadSettingsFragment.kt]: I've refactor this method: [b72405c](https://github.com/Automattic/pocket-casts-android/pull/3166/commits/b72405c68c6ec72db851f1ade7bd04315446ea26)

**Review by geekygecko — APPROVED** [review]

File: modules/features/settings/src/main/java/au/com/shiftyjelly/pocketcasts/settings/AutoDownloadSettingsFragment.kt [hunk 2]
```diff
@@ -236,18 +238,28 @@ class AutoDownloadSettingsFragment :
         updateView()
     }
 
-    private fun handleNewEpisodesToggle(isEnabled: Boolean) {
+    private fun handleNewEpisodesToggle(status: Int) {
         lifecycleScope.launch {
-            updateNewEpisodesSwitch(isEnabled)
+            if (status == Podcast.AUTO_DOWNLOAD_OFF) {
```
Comment (mebarbosa) [inline modules/features/settings/src/main/java/au/com/shiftyjelly/pocketcasts/settings/AutoDownloadSettingsFragment.kt:243]: We want to disable all auto download podcast status when the user manually disabled the New Episodes toggle

**Comment by mebarbosa** [comment]
> So when my app updated the new toggle had switched it off, luckily I caught it and turned my new episodes back on that saved my 400+ new episodes selected, so while yes I would like it to wipe, a backup or delay functionality would be appreciated just in case 😅

Thanks for reporting this @CookieyedCodes ! I've pushed an update that will cover this scenario, so we will keep the previous new episodes toggle status

**Comment by mebarbosa** [comment]
@geekygecko I know you already approved this PR, but would you mind to test it again? 

I had to do one more refactor to cover all scenarios. I already updated the PR description.

File: modules/features/settings/src/main/java/au/com/shiftyjelly/pocketcasts/settings/viewmodel/AutoDownloadSettingsViewModel.kt [hunk 3]
```diff
@@ -24,6 +28,15 @@ class AutoDownloadSettingsViewModel @Inject constructor(
     override val coroutineContext = Dispatchers.Default
     private var isFragmentChangingConfigurations: Boolean = false
 
+    private var _hasEpisodesWithAutoDownloadEnabled: MutableStateFlow<Boolean> = MutableStateFlow(false)
+    val hasEpisodesWithAutoDownloadEnabled: StateFlow<Boolean> = _hasEpisodesWithAutoDownloadEnabled
+
+    init {
+        viewModelScope.launch(Dispatchers.IO) {
+            _hasEpisodesWithAutoDownloadEnabled.value = podcastManager.hasEpisodesWithAutoDownloadStatus(Podcast.AUTO_DOWNLOAD_NEW_EPISODES)
```
Comment (geekygecko) [inline modules/features/settings/src/main/java/au/com/shiftyjelly/pocketcasts/settings/viewmodel/AutoDownloadSettingsViewModel.kt:36]: Is there a chance that this value would be set after it's used in `AutoDownloadSettingsFragment.setupNewEpisodesToggle`?

File: modules/features/settings/src/main/java/au/com/shiftyjelly/pocketcasts/settings/AutoDownloadSettingsFragment.kt [hunk 4]
```diff
@@ -379,6 +391,27 @@ class AutoDownloadSettingsFragment :
             )
     }
 
+    @SuppressLint("CheckResult")
```
[the remark this answers is not in the document]
Comment (mebarbosa) [reply modules/features/settings/src/main/java/au/com/shiftyjelly/pocketcasts/settings/AutoDownloadSettingsFragment.kt:394]: Good catch. Not after I did the second refactor

File: modules/services/preferences/src/main/java/au/com/shiftyjelly/pocketcasts/preferences/Settings.kt [hunk 5]
```diff
@@ -332,7 +334,7 @@ interface Settings {
     val autoDownloadUnmeteredOnly: UserSetting<Boolean>
     val autoDownloadOnlyWhenCharging: UserSetting<Boolean>
     val autoDownloadUpNext: UserSetting<Boolean>
-    val autoDownloadNewEpisodes: UserSetting<Boolean>
+    val autoDownloadNewEpisodes: UserSetting<Int>
```
[the remark this answers is not in the document]
Comment (mebarbosa) [reply modules/services/preferences/src/main/java/au/com/shiftyjelly/pocketcasts/preferences/Settings.kt:337]: I've updated the name of the setting like we talked and it worked for the migration

**Comment by mebarbosa** [comment]
@geekygecko The PR is ready for you to test again. I've addressed the requested changes

**Review by geekygecko — APPROVED** [review]

**Pull request merged**